### DISCUSSION TITLE
New PEG specials

### DISCUSSION THREAD

**Discussion by ianthehenry** [janet-lang/janet#1345] [Ideas]
This is a follow-up to the discussion in https://github.com/janet-lang/janet/pull/1344 -- I was going to write this as a comment but decided it made more sense outside of that PR.

If `sub` is accepted, the other two combinators that have similar behavior to `sub` (in that they also narrow the input string) that I think make sense in addition are `split` and `til`.

# `split`

`split` works just like `string/split`, except that the pattern can be a PEG.

I had originally proposed `sep`, which is a bit more lenient than `split`, as it doesn't have to consume the entire rest of the input. I was trying to make something a little more general, but every time that I've actually wanted something like `sep`, I have really just wanted `split`. And if I ever did want a partial match, I could limit the window of the split with `(sub ... (split ...))`. I think the semantics of `split` are clear and unambiguous, while `sep` involved some decisions that could go either way.

[Implementated here.](https://github.com/ianthehenry/janet/commit/b4db22f9a23b8aea3131c39eccd3c92f14d1a038)

# `til`

`(til term patt)` is similar to `(sequence (sub (to term) patt) (drop term))`, except that it doesn't evaluate the `term` pattern again after the `sub`. I think of it as a combination of `to` and `thru`. This is useful in cases like this:

```janet
(* "- [" (til "]" '(to -1)))
```

Which is something I wrote to parse GHFM task lists like `- [x]`. Currently that requires repeating the terminating `]`, and being careful that the inner pattern doesn't consume the closing `]`.

Combined with a repetition operator like `(some)` or `(any)`, this can behave like split that allows for an optional trailing separator.

I think of it as "capture like `to`, advance like `thru`". `til` was the best name I could come up with but open to better alternatives.

[Implemented here.](https://github.com/ianthehenry/janet/commit/1e365eecec4d25620dc1ae8ccd1ee1f26385984f)

---

I think `split` is pretty useful. I'm not completely confident that `til` is pulling its weight, but I do often write this `(* (to ";") ";")` thing and I find it annoying to repeat the terminator in those cases.

**Comment by pepe** [comment]
I appreciate the thought you have been giving to the PEG lately, especially as it arises from your implementation needs, if I understand it right.
The window (pun intended) you opened with the idea of `sub` makes a lot of sense to me. I would enjoy having `til` merged.
`split` is PEG's missing part, in my opinion.

**Comment by sogaiu** [comment]
For `til`, I struggled a bit to understand the example as-is, but I tried rewriting:

```janet
(* "- [" (til "]" '(to -1)))
```

as:

```janet
(sequence "- ["
          (til "]" (capture (to -1))))
#              ===
#                  -----------------
#                           ^^^^^^^
#                               @@
```

So if we have `(til term patt)`:

* `term` is `"]"`
* `patt` is `(capture (to -1))`

In this case I guess the `-1` refers to the end of what `term` matched and `(to -1)` therefore refers to "whatever was between `- [` and `]`.

Does that seem like an appropriate understanding?

---

For some reason, I found the explanation and code from [the test](https://github.com/ianthehenry/janet/commit/1e365eecec4d25620dc1ae8ccd1ee1f26385984f#diff-b0f6edf26c93104c5a2ee0a7ee9aec27d463462240ceaaa4d189bce0f5531dddR755-R758):

```janet
(test "til: find a separator, match before it, then advance past it"
  ~(* (til "=" '(to -1)) '(to -1))
  "word=something"
  @["word" "something"])
```

easier to comprehend, though I had to think a bit about what the second `'(to -1)` was referring to -- I presume it's supposed to start matching after the `=` and continue to the end of the overall input (and capture).

Don't mind me though, it usually takes me longer to understand things than many others (^^;

**Reply by ianthehenry** [reply]
Yeah, that's exactly right. `'(to -1)` is just short for "capture the rest of the input", which can be narrowed by `sub`/`split`/`til` rules.

**Reply by sogaiu** [reply]
I've thought a bit more and have found the following expression to be easier to understand (just made it 2 lines instead of 1 and indented):

```janet
~(* (til "=" '(to -1))
    '(to -1))
```

With this, I can take in the flow at a glance.  Something like:

* there's a sequence that's got 2 parts
* the first part is up "until" (though including) the first encountered `=`
* the second part starts after the first part (i.e. right after the previously located `=`)

When it was all on one line (and perhaps that there are two instances of `'(to -1)`), this view was not surfacing so readily (^^;

I think I have a better sense of what to scan for when reading constructs where `til` is used in combination with other things.

**Comment by sogaiu** [comment]
I went over the [tests for `split`](https://github.com/ianthehenry/janet/commit/b4db22f9a23b8aea3131c39eccd3c92f14d1a038#diff-b0f6edf26c93104c5a2ee0a7ee9aec27d463462240ceaaa4d189bce0f5531dddR714-R752) and think they are quite helpful in gaining familiarity.

I'm posting them below along with a special note for folks to check out the last one: `~(split :s+ '(to -1))`.

```janet
(test "split: basic functionality"
  ~(split "," '1)
  "a,b,c"
  @["a" "b" "c"])

(test "split: drops captures from separator pattern"
  ~(split '"," '1)
  "a,b,c"
  @["a" "b" "c"])

(test "split: can match empty subpatterns"
  ~(split "," ':w*)
  ",a,,bar,,,c,,"
  @["" "a" "" "bar" "" "" "c" "" ""])

(test "split: subpattern is limited to only text before the separator"
  ~(split "," '(to -1))
  "a,,bar,c"
  @["a" "" "bar" "c"])

(test "split: fails if any subpattern fails"
  ~(split "," '"a")
  "a,a,b"
  nil)

(test "split: separator does not have to match anything"
  ~(split "x" '(to -1))
  "a,a,b"
  @["a,a,b"])

(test "split: always consumes entire input"
  ~(split 1 '"")
  "abc"
  @["" "" "" ""])

(test "split: separator can be an arbitrary PEG"
  ~(split :s+ '(to -1))
  "a   b      c"
  @["a" "b" "c"])
```
